src/backtrace.rs

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturnAddress {
+    ReturnAddress(u64),
+    End,
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::End => write!(f, ""),
+        }
+    }
+}
+
+pub(crate) fn backtrace() -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    unsafe { backtrace_internal(fp) }
+}
+
+// Safety
+// The caller must ensure that fp is a valid address on the stack frame.
+unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
+    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
+    let mut index = 0;
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
```

Comment:
I still think that using a &[u64] slice is much better approach than trying to shift bits and check for boundary conditions. Using a slice only requires _stack_end and _stack_start to be right (and it's a given implicit assumption in both cases). However, when using slices, reviewers (and in the future maintainers) don't have to worry about getting anything else right  in the unsafe blocks (since there are none), and all checks are enforced by the compiler.

Reply:
And with the slice approach you can even iterate over it rather than the current `loop {}` with the `offset` calculation which is more of an index

Reply:
Furthermore, it will make it possible to hoist the skip logic by simply returning the slice, and then using slice.iter().skip(2) at the caller site, which simplifies the code even further.

Reply:
No, we don't need to create a stack iterator because you get it for free after creating a slice. For example, I think you can simply return `[BTReturnAddress; MAX_BACKTRACE_ADDRESSES]` and then use iter() in the return value.

For example:

`
let stack_trace = unsafe {backtrace_internal(fp)};
stack_trace.to_iter().skip(2).for_each(|entry_type|) {
   // Handle entry type.
}
`

Also, notice how we use unsafe{} exactly once below with the linker defined symbols, and then rely on the compiler to enforce the rest without having to worry about whether we got it right.

/```
/ Safety
// The caller must ensure that fp is a valid address on the stack frame.
unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
    let mut index = 0;
    let mut skip = 2;

    // Safe because we trust the linker placed these symbols correctly.
    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
    let mut fp  = fp;
    let stack = unsafe {core::slice::from_raw_parts(stack_start as *const u64, (stack_end - stack_start) as usize)};
    loop {  
        if fp < (stack_start+2) || fp > stack_end {
            break;
        }

        let offset = ((fp - stack_start - 2)/8) as usize;
        if offset > 0 && let Some(address) = stack.get(offset) && let Some(current_fp) = stack.get(offset-1) {
            fp = *current_fp;
            let address = *address;     
            if address == 0 {
                break;
            }
            
            result[index] = BTReturnAddress::ReturnAddress(address);
            index += 1;
    
            if index >= MAX_BACKTRACE_ADDRESSES {
                break;
            }        
        } else {
            break;
        }
    }

    result
}
```